### PR TITLE
[conflict] Fix #1183

### LINKED ISSUE #1183: Incompatible dictionary results in confusing error message
Dictionary literals with incompatible keys/values result in weird error message, because mypy translate the expression internally to something quite different before type checking.

```
from typing import Dict
d = {1 : ''} # type: Dict[int, int]
```

This gives a confusing error message:

```
t2.py:2: error: List item 0 has incompatible type "Tuple[int, str]"
```

### PR DESCRIPTION
**Pull request by fabianhjr**
This is a fix/improvement for #1183

### CODE REVIEW

File: mypy/messages.py [hunk 1]
```diff
@@ -460,11 +460,26 @@ def incompatible_argument(self, n: int, m: int, callee: CallableType, arg_type:
             target = 'to {} '.format(name)
 
         msg = ''
+        try:
+            expected_type = callee.arg_types[m - 1]
+        except IndexError:  # Varargs callees
```
Comment (ddfisher) [inline mypy/messages.py]: It's better to not use exceptions for normal code flow.

File: mypy/messages.py [hunk 2]
```diff
@@ -460,11 +460,26 @@ def incompatible_argument(self, n: int, m: int, callee: CallableType, arg_type:
...
         if callee.name == '<list>':
             name = callee.name[1:-1]
             n -= 1
-            msg = '{} item {} has incompatible type {}'.format(
-                name[0].upper() + name[1:], n, self.format_simple(arg_type))
+            msg = '{} item {} has incompatible type {}; expected {}'.format(
+                name[0].upper() + name[1:], n,
+                self.format_simple(arg_type), self.format_simple(expected_type))
+            # Will add a note if dealing with List[Tuple[K, V]] which might be a Dict[K, V]
+            # TODO: The following is a necesarry condition for a Dict;
```
Comment (ddfisher) [inline mypy/messages.py]: Nit: "necessary"

File: mypy/messages.py (hunk 1 above)
Comment (fabianhjr) [inline mypy/messages.py]: I only moved that existing code up; if it is an issue it could be fix in another PR/commit.

File: mypy/messages.py (hunk 2 above)
Comment (fabianhjr) [inline mypy/messages.py]: Fixed

**Comment by ddfisher** [comment]
I don't have time to give this a detailed review right now, unfortunately -- I'll try to circle back to this in the next couple weeks.

**Comment by fabianhjr** [comment]
Ok thanks, mention me if you need anything. C:

**Comment by gvanrossum** [comment]
It looks like this PR is pretty much stuck. I'm trying to clean up stuck PRs so that the list of PRs is not too long. If you feel like working on this some more, just go ahead and reopen it, or open a new PR -- the corresponding issue will remain open with a reference to this (closed) PR in it, so if someone else feels like tackling this problem they still have this PR as a starting point.

**Comment by fabianhjr** [comment]
@gvanrossum don't worry, its ok. I have been stuck with school. I could take another shot in December. C:

**Comment by gvanrossum** [comment]
OK, see you then!
